[Federation] Add a SchedulingAdapter that can extend the FederatedTypeAdapter and that provides hooks for scheduling objects into clusters.

perotinus opened the pull request:
**Release note**:
```release-note
NONE
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45563)
<!-- Reviewable:end -->

perotinus commented:
cc @marun @irfanurrehman

marun commented:
@k8s-bot ok to test

marun commented:
Good progress!  I think I'd like to see this work follow #45374's example of keeping functions small enough to be unit tested.  It may be worth considering rebasing on that work to ensure that reconcile with the proposed hook methods remains testable.

---

Reviewed 12 of 13 files at r1, 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[federation/pkg/federatedtypes/adapter.go, line 58 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjXh7wsnUkdaV8s0rh:-KjjXh7wsnUkdaV8s0ri:bnc0qp4) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federatedtypes/adapter.go#L58)):*
> ```Go
> 	ClusterWatch(client kubeclientset.Interface, namespace string, options metav1.ListOptions) (watch.Interface, error)
> 
> 	// The following operations allow the adapter to modify objects before they are propogated
> ```

Consider following the approach suggested in https://github.com/kubernetes/kubernetes/pull/45497#discussion_r115620797 to minimize unnecessary no-ops in adapters that don't need these functions.

---

*[federation/pkg/federatedtypes/replicaset.go, line 95 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjayT6PTZhf1iMsmKg:-KjjayT6PTZhf1iMsmKh:ba3dfna) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federatedtypes/replicaset.go#L95)):*
> ```Go
> }
> 
> //		lrs := &extensionsv1.ReplicaSet{
> ```

Remove?

---

*[federation/pkg/federatedtypes/replicaset.go, line 172 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-Kjjdcn1VgVYYWCuxkft:-Kjjdcn1VgVYYWCuxkfu:bik9mtp) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federatedtypes/replicaset.go#L172)):*
> ```Go
> }
> 
> func (a *ReplicaSetAdapter) PrepareForUpdateFunc() func(obj pkgruntime.Object, key string, clusters []*fedv1beta1.Cluster, informer fedutil.FederatedInformer) (interface{}, error) {
> ```

I reviewed @irfanurrehman's PR first, which used more specific hook methods instead of the generic events you propose here.  Consider coordinating with him to come to agreement on a mutually acceptable path forward.

---

*[federation/pkg/federatedtypes/replicaset.go, line 173 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjbJtLQASz34SLxDvx:-KjjbJtLQASz34SLxDvy:bp763kc) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federatedtypes/replicaset.go#L173)):*
> ```Go
> 
> func (a *ReplicaSetAdapter) PrepareForUpdateFunc() func(obj pkgruntime.Object, key string, clusters []*fedv1beta1.Cluster, informer fedutil.FederatedInformer) (interface{}, error) {
> 	return func(obj pkgruntime.Object, key string, clusters []*fedv1beta1.Cluster, informer fedutil.FederatedInformer) (interface{}, error) {
> ```

Why is it desirable to return the function vs having the client call invoke it directly?  Same comment for the other hook implementations.

Are you intending to ensure unit test coverage for the hook methods?

---

*[federation/pkg/federation-controller/sync/controller.go, line 335 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjYeHN3Zkjtg6f-BTS:-KjjYeHN3Zkjtg6f-BTT:bo4vyev) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federation-controller/sync/controller.go#L335)):*
> ```Go
> func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName) (reconciliationStatus, error) {
> 	if !s.isSynced() {
> 		return statusNotSynced, nil
> ```

As per our offline discussion, I think it is preferable to minimize the scope of any one function such that logging can be done inline and testing can still reliably validate failure modes without having to look at error text.

---

*[federation/pkg/federation-controller/sync/controller.go, line 473 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjZWK4gB72CQTeDCn8:-KjjZWK4gB72CQTeDCn9:bka6gam) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federation-controller/sync/controller.go#L473)):*
> ```Go
> 	}
> 
> 	// Evertyhing is in order but let's be double sure
> ```

I switched to returning 'ok' in #45374.  Unless I'm missing something, 'double sure' won't do anything but mask potential errors.

---

*[federation/pkg/federation-controller/sync/replicasetcontroller_test.go, line 47 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-Kjj_VEN3QVH_iwMXJvN:-Kjj_VEN3QVH_iwMXJvO:b-f9bupf) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federation-controller/sync/replicasetcontroller_test.go#L47)):*
> ```Go
> )
> 
> //func TestParseFederationReplicaSetReference(t *testing.T) {
> ```

Should this be removed?  I'm guessing #44525 ensures this anyway, consider rebasing on top of that.

---

*[federation/pkg/federation-controller/util/podanalyzer/pod_helper.go, line 20 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjaEbdlVPGjOzeF2Pu:-KjjaEbdlVPGjOzeF2Pv:bb48dgp) ([raw file](https://github.com/kubernetes/kubernetes/blob/e537cec9b0b39a4d5f503e8a35706d726deb9b7d/federation/pkg/federation-controller/util/podanalyzer/pod_helper.go#L20)):*
> ```Go
> 
> import (
> 	"fmt"
> ```

It looks like the changes to this package impact the deployment controller, but there are no changes to that controller included in this PR.

Consider submitting the pod changes in a separate PR to simplify life for your reviewers. 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-:-KjjeVL88STu7Acve4KE:bnaixue)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 5 of 22 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[federation/pkg/federatedtypes/adapter.go, line 58 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjXh7wsnUkdaV8s0rh:-KkShW7-OJ-p5vBlNmsj:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federatedtypes/adapter.go#L58)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider following the approach suggested in https://github.com/kubernetes/kubernetes/pull/45497#discussion_r115620797 to minimize unnecessary no-ops in adapters that don't need these functions.
</blockquote></details>

Done.

---

*[federation/pkg/federatedtypes/replicaset.go, line 95 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjayT6PTZhf1iMsmKg:-KkShWSHCagGazWgz_2M:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federatedtypes/replicaset.go#L95)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Remove?
</blockquote></details>

Done.

---

*[federation/pkg/federatedtypes/replicaset.go, line 173 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjbJtLQASz34SLxDvx:-KkShY9azVLb_iqww-fg:b-uve6ni) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federatedtypes/replicaset.go#L173)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why is it desirable to return the function vs having the client call invoke it directly?  Same comment for the other hook implementations.

Are you intending to ensure unit test coverage for the hook methods?
</blockquote></details>

I've modified this 

I have to think a bit more about test coverage for the hook methods: I agree that it makes sense and should be done, but passing in the federated informer will make these tests a bit more ugly. Perhaps the solution here is to have the hooks be lightweight wrappers that call into other methods.

---

*[federation/pkg/federation-controller/sync/controller.go, line 335 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjYeHN3Zkjtg6f-BTS:-KkShyJuboX8zt4xxpzl:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federation-controller/sync/controller.go#L335)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

As per our offline discussion, I think it is preferable to minimize the scope of any one function such that logging can be done inline and testing can still reliably validate failure modes without having to look at error text.
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/sync/controller.go, line 473 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjZWK4gB72CQTeDCn8:-KkShy3KP5KZTwS6l4U4:b7jiio3) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federation-controller/sync/controller.go#L473)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I switched to returning 'ok' in #45374.  Unless I'm missing something, 'double sure' won't do anything but mask potential errors.
</blockquote></details>

I rebased on top of your PR, so this should be good now.

---

*[federation/pkg/federation-controller/sync/replicasetcontroller_test.go, line 47 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-Kjj_VEN3QVH_iwMXJvN:-KkSi3hhczbQZWpArxh1:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federation-controller/sync/replicasetcontroller_test.go#L47)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Should this be removed?  I'm guessing #44525 ensures this anyway, consider rebasing on top of that.
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/util/podanalyzer/pod_helper.go, line 20 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjaEbdlVPGjOzeF2Pu:-KkSi5_HzYsMix_0KnWH:bnlvu2i) ([raw file](https://github.com/kubernetes/kubernetes/blob/e537cec9b0b39a4d5f503e8a35706d726deb9b7d/federation/pkg/federation-controller/util/podanalyzer/pod_helper.go#L20)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

It looks like the changes to this package impact the deployment controller, but there are no changes to that controller included in this PR.

Consider submitting the pod changes in a separate PR to simplify life for your reviewers. 
</blockquote></details>

These changes should be isolated to the first commit which is being reviewed in https://github.com/kubernetes/kubernetes/pull/45252. However, that PR may be dropped because of testing issues, in which case the code to modify the pod analysis method will potentially be moved here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563)*
<!-- Sent from Reviewable.io -->

perotinus commented:
@k8s-bot test this

irfanurrehman commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -438,10 +450,32 @@ func syncToClusters(clustersAccessor clustersAccessorFunc, operationsAccessor op
-	operations, err := operationsAccessor(adapter, clusters, obj)
+	var userInfo interface{}
+	if objUpdatingAdapter, ok := adapter.(federatedtypes.ObjectUpdatingAdapter); ok {
+		userInfo, err = objUpdatingAdapter.PrepareForUpdate(obj, key, clusters, informer)
```
I see one problem here when used with cluster selector.
The schedule is called before the cluster selector logic over all clusters in the list and then those clusters(with the result) are discarded which do not get selected by cluster selector. The over all schedule is thus wrong. The overall schedule logic should work on only those clusters which get selected by cluster selector.

perotinus replied:
I believe that my work in #46252 will address this: the cluster selection happens before any of the scheduling, which is what I believe is supposed to happen.

irfanurrehman commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -452,22 +486,23 @@ func syncToClusters(clustersAccessor clustersAccessorFunc, operationsAccessor op
+		desiredObj := adapter.Copy(obj)
+		objMeta := adapter.ObjectMeta(desiredObj)
+
 		clusterObj, found, err := accessor(cluster.Name)
```
While in the process of rebasing my work on this, I met with another problem...
In my scheduling algorithm, I need to make decisions as to which cluster gets an object (even if its not there), which cluster gets the object updated, and also remove object from some cluster if needed (even if it exists).
In this flow, the same cluster objects gets accessed 2 different places (at one place in PrepareForUpdate()using informer directly, and another one here...) If I try to keep this same flow, and try to use just accessed local object(at this line of comment), there is a possible race condition for me..(i get an object here, which wasnt present when I accessed in prepareforupdate, my earlier schedule decision is wrong). On the other hand if I stick to the schedule() computation made in PrepareForUpdate() (which tells me ab all three conditions - to remove object, to add object, to update object) already and ignore the object accessed here, I need to introduce some crooked logic only for this hpa adapter, which doesn't sound right to me......

So my need and use case still is the same, and kind of the super set of what is needed for replicasets (the rs controller never deletes an object from any cluster in reconcile) and the simple steps that I want to achieve are still same.
- get the list of clusters with the objects (with nil or some-other way to indicate if no object in that cluster).
- schedule makes decisions where to add, where to remove and where to update, conveys the result back.
- clusteroperations makes a list of operations to act on that decision.

Is there some way, we can update this logic to achieve above.

Additional note to this, which somehow I could not express properly in our discussion yesterday, was about the methods. Actually its a wrong term we are using. PrepareForUpdate( ) is an interface first (then a method on a structure), and our goal should be to keep the interfaces to minimum, and I sort of want to stress my opinion of keeping these interfaces to minimum. Our goal should also be to keep the information exchange to and from the interfaces to minimum (userstate here) especially if its not needed for any controller back in reconcile as of now (YAGNI). Implementers can choose to structure/break their code in as many methods/functions internally for the interface implementation. 

With that said, I am totally ready to rebase on top of this, if everybody thinks this is the right approach.
I just want this to reach to some conclusion, ASAP.. :-)
( I wrote all this because, I tried to rebase on top, this second time, and got stuck with no direct solution.. :-( )

irfanurrehman replied:
cc @marun

perotinus replied:
Hmm. I think I see your point. One potential way we could address this is by getting the clusters once in a loop at the beginning of reconcile() and using that list of clusters throughout the method. Would that solve your issue? The entire reconcile loop would use the same cluster objects in that case.

I think your point about having a minimal interface is valid, but I don't think that the smallest number of methods or amount of data passed around is necessarily the most minimal interface. Having one method in an interface which has semantic complexity and is hard to understand is less "minimal" than having more methods that are more like a unit.

Also, I think I forgot to repush this branch: I changed these methods quite a bit to be more scheduling specific. I'll do that now.

irfanurrehman commented:
@perotinus, I suggest you separate the additional adapter logic in a separate PR as I did in https://github.com/kubernetes/kubernetes/pull/45497.
Also, plz check the review comments from @marun (especially about introducing IsSchedule() method on normal sync adapter, u can choose a suitable name) and ensure that they are taken care here.
It will be easier for review and for others (me.. :) ) to base their work on top.
Thanks for all the hard work!!

perotinus committed:
[Federation] Add a SchedulingAdapter that can extend the FederatedTypeAdapter and that provides hooks for scheduling objects into clusters.

irfanurrehman commented on federation/pkg/federatedtypes/adapter.go:
```diff
@@ -53,9 +55,35 @@ type FederatedTypeAdapter interface {
 
+// SchedulingStatus contains the status of the objects that are being
+// scheduled into joined clusters.
+type SchedulingStatus struct {
```
Please update these to generic types (or make them interfaces, where they are passed/used).

irfanurrehman commented on federation/pkg/federatedtypes/adapter.go:
```diff
@@ -53,9 +55,35 @@ type FederatedTypeAdapter interface {
+
+// SchedulingInfo wraps the information that a SchedulingAdapter needs
+// to update objects per a schedule.
+type SchedulingInfo struct {
```
as above, hpa has different set of properties to pass around!

irfanurrehman commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -444,10 +445,38 @@ func syncToClusters(clustersAccessor clustersAccessorFunc, operationsAccessor op
+			glog.Fatalf("Adapter for kind %q does not properly implement SchedulingAdapter.", kind)
+			return statusError
+		}
+		schedulingInfo, err = schedulingAdapter.GetSchedule(obj, key, selectedClusters, informer)
```
Please ensure that, the cluster objects are queried and passed as one set here (probably as a map), indicating which cluster has the object, and which does not (one way is use a map of same size as current available clusters with nil values identifying absence of object in that cluster).
Other option I could think of is;  don't query the objects here at all; just pass the informer here, and let this interface or any other one in the flow later return the existing set vs scheduled set of objects (or only scheduled set of objects indicating, in which cluster to update, into which cluster to delete and which cluster to create ... (Or else I can't think of how exactly to do it ).
Probably u can think of on these lines:
say in the rs controller schedule, as a result of schedule, the local cluster rs can be deleted from some clusters also
so there has to be some way for this reconcile to be able to finalise the result comparing the original set of object vs new set of objects including all three conditions:
- if not exist in a cluster in existing set,  schedule says create it.
- if exist in a cluster in existing set, schedule says update it.
- if exist in a cluster in existing set, schedule says delete it.

Also the above to be done without race conditions (means objects or set of cluster objects queried at one location to come up with the schedule result)

irfanurrehman reviewed (changes requested):
Hope u can take care of these.
thanks in advance!

marun commented:
While I think further refinement is possible, code freeze is looming and I think supporting replicasets with the sync controller represents is a necessary evolution towards a more maintainable codebase.  Modulo the minor concerns inline, and the need for a test-specific equivalency check on the adapter interface, I'll be ready to lgtm.

---

Reviewed 7 of 20 files at r4, 7 of 22 files at r6, 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[federation/pkg/federatedtypes/adapter.go, line 63 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlLZRu3nEaOKWdF0KeR:-KlLZRu3nEaOKWdF0KeS:boixszd) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federatedtypes/adapter.go#L63)):*
> ```Go
> }
> 
> // SchedulingStatus contains the status of the objects that are being
> ```

(No action required) Would it make sense to put the scheduling-specific stuff into its own file?

---

*[federation/pkg/federatedtypes/adapter.go, line 65 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjEAS-xsv2E4j2KVvl-r7-65:-KlLY1BAgRODnqjzGlo4:b-5lo5gd) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federatedtypes/adapter.go#L65)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Please update these to generic types (or make them interfaces, where they are passed/used).
</blockquote></details>

I think it would be preferable to merge this very specific proposal and have your follow-on make it generic.

---

*[federation/pkg/federation-controller/sync/controller.go, line 449 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOhFHklY1QE8Okc0t0:-KlOhFHklY1QE8Okc0t1:boy54im) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L449)):*
> ```Go
> 
> 	var schedulingInfo *federatedtypes.SchedulingInfo
> 	if adapter.IsSchedulingAdapter() {
> ```

(No action required) Consider simplifying as follows:

```golang
schedulingInfo, err := getSchedulingInfo(adapter, obj, key, selectedClusters, informer)
if err {
  return statusError
}
```

---

*[federation/pkg/federation-controller/sync/controller.go, line 453 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOmmhd8_OQFI19J-uK:-KlOmmhd8_OQFI19J-uL:b-lneu64) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L453)):*
> ```Go
> 		if !ok {
> 			glog.Fatalf("Adapter for kind %q does not properly implement SchedulingAdapter.", kind)
> 			return statusError
> ```

Is this return statement necessary given that ``glog.Fatalf`` calls ``os.Exit``?

---

*[federation/pkg/federation-controller/sync/controller.go, line 470 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOgiTD5aiEv-rmNYgK:-KlOgiTD5aiEv-rmNYgL:b-7ropln) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L470)):*
> ```Go
> 		schedulingAdapter, ok := adapter.(federatedtypes.SchedulingAdapter)
> 		if !ok {
> 			glog.Fatalf("Adapter for kind %q does not properly implement SchedulingAdapter.", kind)
> ```

ditto

---

*[federation/pkg/federation-controller/sync/controller.go, line 522 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOn8A-jZHgJ6S5vOoL:-KlOn8A-jZHgJ6S5vOoM:b-jww1mc) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L522)):*
> ```Go
> 		// The data should not be modified.
> 		desiredObj := adapter.Copy(obj)
> 		kind := adapter.Kind()
> ```

Saving a copy of the kind could be done outside of the loop.

---

*[federation/pkg/federation-controller/sync/controller.go, line 537 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOoW9L4TY6jiV3XkwG:-KlOoW9L4TY6jiV3XkwH:baajjrw) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L537)):*
> ```Go
> 				err = fmt.Errorf("adapter for kind %s does not properly implement SchedulingAdapter.", kind)
> 				glog.Fatalf("Error: %v", err)
> 				return nil, err
> ```

Again, not sure what the point is of returning past ``glog.Fatalf``.

---

*[federation/pkg/federation-controller/sync/controller.go, line 570 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOqBOuTtiKUSYVDXUP:-KlOqBOuTtiKUSYVDXUQ:b36zl1s) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L570)):*
> ```Go
> 	}
> 
> 	for _, cluster := range unselectedClusters {
> ```

Why is it desirable to separating clusters into a slice for each of selected and unselected?  Did you consider using a mapping of cluster name to a boolean indicating selection status?

If separate iteration is preferable, consider breaking this functiondecomposing this function to simplify testing:

```golang
operations := make([]util.FederatedOperation, 0)
getOperationsForSelectedClusters(operations, ...)
getOperationsForUnselectedClusters(operations, ...)
```

---

*[federation/pkg/federation-controller/sync/controller.go, line 573 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOqTt9v3fJWlk9k5IP:-KlOqTt9v3fJWlk9k5IQ:bykqajl) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L573)):*
> ```Go
> 		clusterObj, found, err := accessor(cluster.Name)
> 		if err != nil {
> 			wrappedErr := fmt.Errorf("Failed to get %s %q from cluster %q: %v", adapter.Kind(), key, cluster.Name, err)
> ```

s/adapter.Kind()/kind/

---

*[federation/pkg/federation-controller/sync/controller_test.go, line 163 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOpYJA8q1HMZeJprHT:-KlOpYJA8q1HMZeJprHU:b-pq0sy6) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller_test.go#L163)):*
> ```Go
> }
> 
> func TestClusterOperations(t *testing.T) {
> ```

Are there more test changes pending?  The changes to ``clusterOperations`` suggest a need for more coverage than is added here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-:-KlPGbe5mDy3KRIrihF9:bf1ithh)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 6 of 7 files reviewed at latest revision, 15 unresolved discussions.

---

*[federation/pkg/federatedtypes/adapter.go, line 63 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlLZRu3nEaOKWdF0KeR:-KlPgfU-f1cMumdHUBsg:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federatedtypes/adapter.go#L63)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Would it make sense to put the scheduling-specific stuff into its own file?
</blockquote></details>

Done.

---

*[federation/pkg/federatedtypes/adapter.go, line 65 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjEAS-xsv2E4j2KVvl-r7-65:-KlPggAqN5ApqZI4ad7l:bjovmz7) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federatedtypes/adapter.go#L65)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I think it would be preferable to merge this very specific proposal and have your follow-on make it generic.
</blockquote></details>

I'm inclined to agree: I don't understand the HPA use case as fully as you do, so my efforts to make this generic would probably not be as effective and clean as yours.

---

*[federation/pkg/federatedtypes/adapter.go, line 74 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjEAS-xsv2E4j2KVvl-r7-74:-KlPgqpNpGDZY_XP0taH:bvivaqm) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federatedtypes/adapter.go#L74)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

as above, hpa has different set of properties to pass around!
</blockquote></details>

That makes sense. In light of Maru's comment above, I think I'd prefer to defer this work to you, since you understand your use case better.

---

*[federation/pkg/federation-controller/sync/controller.go, line 449 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOhFHklY1QE8Okc0t0:-KlPlUTXy1oiZPpzwUeD:b-z0h6bd) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L449)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Consider simplifying as follows:

```golang
schedulingInfo, err := getSchedulingInfo(adapter, obj, key, selectedClusters, informer)
if err {
  return statusError
}
```
</blockquote></details>

Do you mean extract a method that gets the scheduling info and hides the logic for casting and checking for isScheduleObject?  I can do that, but it seems a bit odd to extract just this one scheduling method call and leave the other two inline below.

---

*[federation/pkg/federation-controller/sync/controller.go, line 453 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOmmhd8_OQFI19J-uK:-KlPkGqRT57G1gCGaRBe:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L453)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Is this return statement necessary given that ``glog.Fatalf`` calls ``os.Exit``?
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/sync/controller.go, line 455 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KjjEAS-xsv2E4j2KVvs-r7-455:-KlPkUDG7SVb0z7TQ_hm:b-6vyomk) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L455)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Please ensure that, the cluster objects are queried and passed as one set here (probably as a map), indicating which cluster has the object, and which does not (one way is use a map of same size as current available clusters with nil values identifying absence of object in that cluster).
Other option I could think of is;  don't query the objects here at all; just pass the informer here (as you are doing), and let this interface or any other one in the flow later return the existing set vs scheduled set of objects (or only scheduled set of objects indicating, in which cluster to update, into which cluster to delete and which cluster to create ... (Or else I can't think of how exactly to do it ).
Probably u can think of on these lines:
say in the rs controller schedule, as a result of schedule, the local cluster rs can be deleted from some clusters also
so there has to be some way for this reconcile to be able to finalise the result comparing the original set of object vs new set of objects including all three conditions:
- if not exist in a cluster in existing set,  schedule says create it.
- if exist in a cluster in existing set, schedule says update it.
- if exist in a cluster in existing set, schedule says delete it.

Also the above to be done without race conditions (means objects or set of cluster objects queried at one location to come up with the schedule result); and probably then the originally queried set and the newly scheduled set compared to arrive at cluster operations.
</blockquote></details>

I agree that this is a good strategy. In the interest of getting this PR into 1.7 and not blocking, I'd prefer to defer this work.

---

*[federation/pkg/federation-controller/sync/controller.go, line 470 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOgiTD5aiEv-rmNYgK:-KlPkHUKpB-jCS70IU1G:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L470)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/sync/controller.go, line 522 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOn8A-jZHgJ6S5vOoL:-KlPkICSJgGyNH8Q1Mmq:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L522)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Saving a copy of the kind could be done outside of the loop.
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/sync/controller.go, line 537 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOoW9L4TY6jiV3XkwG:-KlPjA-0Ie8mujfmn2Ko:bxc5zoi) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L537)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Again, not sure what the point is of returning past ``glog.Fatalf``.
</blockquote></details>

This is probably an old habit from my Objective-C days, where C assertions were disabled in release-mode code so the pattern was to assert and then return. Removed.

---

*[federation/pkg/federation-controller/sync/controller.go, line 570 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOqBOuTtiKUSYVDXUP:-KlPjOnzHpjnzJi1KNKS:b2zq887) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L570)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why is it desirable to separating clusters into a slice for each of selected and unselected?  Did you consider using a mapping of cluster name to a boolean indicating selection status?

If separate iteration is preferable, consider breaking this functiondecomposing this function to simplify testing:

```golang
operations := make([]util.FederatedOperation, 0)
getOperationsForSelectedClusters(operations, ...)
getOperationsForUnselectedClusters(operations, ...)
```
</blockquote></details>

This seemed simpler in my head. I think it would have been nice to express this with set operations, but in the absence of that, the two slices seemed like a good solution.

Also, not that this code went in in a different PR.

Would you be OK deferring changing this until a later PR? I don't really want to bolt testing onto this PR, I'd rather get this in and then think about how to factor it cleanly for good tests without a looming deadline.

---

*[federation/pkg/federation-controller/sync/controller.go, line 573 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOqTt9v3fJWlk9k5IP:-KlPk2sVyPHQO3UyTj_f:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L573)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

s/adapter.Kind()/kind/
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/sync/controller_test.go, line 163 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOpYJA8q1HMZeJprHT:-KlPkKMiQ5Tu19IGhQ8c:b9j932z) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller_test.go#L163)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Are there more test changes pending?  The changes to ``clusterOperations`` suggest a need for more coverage than is added here.
</blockquote></details>

Yes, I agree. Since the code freeze is looming, are you OK doing more extensive test changes in a follow up?

---

*[federation/pkg/federatedtypes/replicaset.go, line 172 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-Kjjdcn1VgVYYWCuxkft:-KlPk5fncsJoVVJSfntG:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/83ebb8ef5fb76b081e4af487a400eeb038cf7199/federation/pkg/federatedtypes/replicaset.go#L172)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I reviewed @irfanurrehman's PR first, which used more specific hook methods instead of the generic events you propose here.  Consider coordinating with him to come to agreement on a mutually acceptable path forward.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Thanks for the review! I'm working on the test equivalency change now.

---

Review status: 6 of 7 files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-:-KlPlgwB3snb117UAX0A:b-dikaat)*
<!-- Sent from Reviewable.io -->

perotinus commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

perotinus commented:
OK, the test equivalency changes are pushed, and will soon be added to #46527.

marun commented:
/lgtm

---

Reviewed 4 of 4 files at r9.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federation-controller/sync/controller.go, line 449 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOhFHklY1QE8Okc0t0:-KlUlxFEXzcS0-ud1QAP:b-om9rmg) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L449)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Do you mean extract a method that gets the scheduling info and hides the logic for casting and checking for isScheduleObject?  I can do that, but it seems a bit odd to extract just this one scheduling method call and leave the other two inline below.
</blockquote></details>

(No action required) I meant to suggest that you do the same for the other 2 methods as well.  I think it would be cleaner, but it can be done in the future if at all.

---

*[federation/pkg/federation-controller/sync/controller.go, line 570 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOqBOuTtiKUSYVDXUP:-KlUmN9cSZPA92Jf5Vb9:bok6o1k) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L570)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

This seemed simpler in my head. I think it would have been nice to express this with set operations, but in the absence of that, the two slices seemed like a good solution.

Also, not that this code went in in a different PR.

Would you be OK deferring changing this until a later PR? I don't really want to bolt testing onto this PR, I'd rather get this in and then think about how to factor it cleanly for good tests without a looming deadline.
</blockquote></details>

Yeah, it's fine for now.

---

*[federation/pkg/federation-controller/sync/controller_test.go, line 163 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOpYJA8q1HMZeJprHT:-KlUmRB3cQJp4vXaGIhe:br02ght) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller_test.go#L163)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Yes, I agree. Since the code freeze is looming, are you OK doing more extensive test changes in a follow up?
</blockquote></details>

I'm ok with that given that the proposed change shouldn't result in a change in coverage, just that the coverage could stand to be improved.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-:-KlUnR9uQy_PvOman6wZ:b-8ks5qa)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federation-controller/sync/controller.go, line 449 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563#-KlOhFHklY1QE8Okc0t0:-KlV2NO2Z7JYCAJAwG_v:bltw5e4) ([raw file](https://github.com/kubernetes/kubernetes/blob/43235315db4eeafa9c51699dbc50aa97b69a0b50/federation/pkg/federation-controller/sync/controller.go#L449)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) I meant to suggest that you do the same for the other 2 methods as well.  I think it would be cleaner, but it can be done in the future if at all.
</blockquote></details>

OK, SGTM.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45563)*
<!-- Sent from Reviewable.io -->

csbell commented:
/approved
